Plots for TauP raypaths

krischer opened the pull request:
This PR adds some basic plotting capabilities to `obspy.taup`. It offers two different types of plots: `spherical` and `cartesian` ones, it defaults to `spherical`.

Usage:

```python
>>> from obspy.taup import TauPyModel
>>> m = TauPyModel()
>>> m.get_ray_paths(500, 130).plot()
```

![rays](https://cloud.githubusercontent.com/assets/800487/6228480/c7147cf4-b6ac-11e4-9dae-fa44ab0a7de9.png)

```python
>>> m.get_ray_paths(500, 130).plot(plot_type="cartesian")
```

![rays2](https://cloud.githubusercontent.com/assets/800487/6228510/281250ee-b6ad-11e4-8f3d-4d9d0ea806bb.png)

@crotwell: Sometimes the paths don't end at the rights points. Is this a consequence of the model not being sampled fine enough, or some other issue? Two examples:

```python
>>> m.get_ray_paths(10, 5, phase_list=["P"]).plot(plot_type="cartesian", plot_all=False)
```

![problem](https://cloud.githubusercontent.com/assets/800487/6228616/7eb7da44-b6ae-11e4-9769-219d5fac3e53.png)


```python
>>> m.get_ray_paths(0, 10, phase_list=["PPPPPP"]).plot(plot_type="spherical", plot_all=True)
```
![problem2](https://cloud.githubusercontent.com/assets/800487/6228558/9ab67986-b6ad-11e4-9a58-5068fe21b349.png)

megies commented:
Looks great! Legend maybe needs extra room?

krischer commented:
Rewriting the `obspy.taup` docs is next on my list.

The problem with the legend more or less disappears if plotted interactively and one can just resize the window. I also guess that in most real use cases people will plot way less phases.

crotwell commented:
Regarding the end points not lining up, is this something that happens with my java TauP? If so then it is probably a bug and I should look into it. If not, then I guess it is a artifact of the translation. Checking that is probably worth doing in general.

The only other thought I have that might help is have you checked for paths that go the "wrong" way around the earth? For example of PPPPPP, there is a path that goes clockwise and another one that goes counterclockwise. If you don't "flip" the long way around path when plotting it will not end at the correct location. In your example one ray travels 350 degrees to end up 10 degrees away from the start, another travels 370 degrees, starting in the opposite direction, and also ends up at the same end point. THis is the purpose of the "Purist Distance" column in the output of taup_time.

krischer commented:
@crotwell: Thanks for your help with this. I think you are correct regarding the `PPPPPP` plot. I'll fix it. This one also works with your Java TauP.

The issue with the local `P` phases persists in your Java code though. Sorry..the plots are not as pretty. Code to produce them is at the end (your Java TauP version does not appear to be able to plot in a zoomed in fashion).

![untitled](https://cloud.githubusercontent.com/assets/800487/6229272/1207c14a-b6b6-11e4-86a4-3e002e3737b1.png)

![untitled](https://cloud.githubusercontent.com/assets/800487/6229274/179d7730-b6b6-11e4-9525-23778983fb31.png)

```python
import matplotlib.pylab as plt
import numpy as np
import subprocess

taup_path = "/Users/lion/Downloads/TauP-2.1.2/bin/taup_path"

# Local plot.
out = subprocess.Popen(
        [taup_path + " -mod iasp91 -pH 'P' -h 10 -deg 5 -o stdout"],
        shell=True,
        stdout=subprocess.PIPE).communicate()[0].decode()

plt.figure(1)

plt.gca().invert_yaxis()
phases = []
current_phase = None
for line in out.strip().split("\n"):
    line = line.strip()
    if line.startswith(">"):
        if current_phase:
            phases.append(current_phase)
        current_phase = []
        continue
    dist, depth = map(float, line.split())
    current_phase.append((dist, 6371 - depth))
if current_phase:
    phases.append(current_phase)


for phase in phases:
    plt.plot([_i[0] for _i in phase], [_i[1] for _i in phase])


# Spherical PPPPPP plot
out = subprocess.Popen(
        [taup_path + " -mod iasp91 -pH 'PPPPPP' -h 0 -deg 10 -o stdout"],
        shell=True,
        stdout=subprocess.PIPE).communicate()[0].decode()

phases = []
current_phase = None
for line in out.strip().split("\n"):
    line = line.strip()
    if line.startswith(">"):
        if current_phase:
            phases.append(current_phase)
        current_phase = []
        continue
    dist, depth = map(float, line.split())
    current_phase.append((np.deg2rad(dist), depth))
if current_phase:
    phases.append(current_phase)

plt.figure(2)
ax = plt.subplot(111, polar=True)
ax.set_theta_zero_location('N')
ax.set_theta_direction(-1)
ax.set_xticks([])
ax.set_yticks([])

for phase in phases:
    ax.plot([_i[0] for _i in phase], [_i[1] for _i in phase])

ax.xaxis.set_major_formatter(plt.NullFormatter())
ax.yaxis.set_major_formatter(plt.NullFormatter())
ax.set_rmin(0)
ax.set_rmax(6371)

plt.show()
```

krischer commented:
@crotwell I did find some other (a bit obscure) phases where it does not quite add up. The issue is identical in your Java version and our Python port. On the left is always the Python version, on the right the Java version.

![untitled](https://cloud.githubusercontent.com/assets/800487/6229800/abf51eb6-b6ba-11e4-8d3e-b43588f34e41.png)

crotwell commented:
I'll check into these. I suspect this is the case of rounding errors for phases that traverse a lot of layers. It might be that I need to up the sampling in this case. Thanks for letting me know.

QuLogic commented on obspy/taup/tau.py:
```diff
@@ -4,6 +4,12 @@
+
+import matplotlib.cbook
+import matplotlib.pyplot as plt
+import numpy as np
```
NumPy, matplotlib, then cbook.

krischer replied:
This is still a bit confusing to me. I like the general structure future -> stdlib -> third party -> obspy but I think each block in itself should be alphabetically ordered. The numpy -> matplotlib -> scipy order and all at the end of the third party block makes little sense to me...but I value consistency more than this so I guess I'll change it ;-)

QuLogic replied:
For scientific packages, it's SciPy conventions, which unfortunately are not alphabetical.

krischer replied:
Is there any reason why matplotlib is before scipy? Chronologically plotting comes after the analysis ;-)

QuLogic replied:
Probably because I was thinking of SciPy and forgot they'd have put scipy in the first-party location. :stuck_out_tongue:

QuLogic commented on obspy/taup/tau.py:
```diff
@@ -11,14 +17,35 @@
 from .taup_time import TauP_Time
 
 
+# From colorbrewer2.org
```
Which set?

QuLogic replied:
You should be able to get these from matplotlib. From #915:
```
cmap = plt.get_cmap('Paired', lut=6)
colors = ['#%02x%02x%02x' % tuple(col * 255 for col in cmap(i)[:3])
          for i in range(6)]
```

QuLogic commented on obspy/taup/tau.py:
```diff
@@ -31,6 +58,131 @@ def __str__(self):
+            A spherical plot is always global whereas a cartesian one can
+            also be local.
+        :type plot_type: str
+        :param plot_all: By default only rays that match the requested
```
I'm confused by the name here. It's plot _all_, and the description is "by default only something...", but the default is `plot_all=True`, so it's either named the opposite of what it does, or the default is wrong.

krischer replied:
Ah true...i changed the default behavior in some commit and forgot to adopt the doc. thanks.

QuLogic commented on obspy/taup/tau.py:
```diff
@@ -31,6 +58,131 @@ def __str__(self):
+            if abs(dist - self.distance) / dist > 1E-5:
+                if plot_all is False:
+                    continue
+                # Mirror on axis.
```
Should we just do this lower down in the stack to avoid confusion?

krischer replied:
I also thought about this and am not really sure as we'd then end up with negative distances which might confuse even more people.

QuLogic commented:
I was wondering if there might some use to an alternative labelling method, e.g., labelling arrivals at their receiver:
![phases](http://www.staff.uni-bayreuth.de/~bt220235/classes/intro-geophys/figs/phases.jpg)

Say you asked for P at x° and PP at 2x° (you can do that, right?), then the first leg would overlap and you wouldn't see one of them.

This might be a bit tricky if you have many arrivals near each other.

krischer commented:
This would of course be very nice but kind of impossible (or at least very hard to do) as all of our phases have the same receiver as of now. The plotting is trivial to extend to allow arrivals at multiple receivers/distances but it would be pretty hard to actually get nicely separated single arrivals which are needed for your proposed plot.

A plot like this would still be awesome for the tutorial and the gallery but it would probably need to be a one-off plot. The differentiated P and S legs though we should still do :-)

megies commented:
In the cartesian plot, if there are arrivals coming in from the opposite direction, maybe there should be a second receiver triangle at `-= 360` degrees?

![](https://cloud.githubusercontent.com/assets/800487/6228510/281250ee-b6ad-11e4-8f3d-4d9d0ea806bb.png)

QuLogic commented:
I added some simple plots to the tutorial as well as one approximating the above. I tried to fit in complicated phases like `SKPPKP`, but contrary to that above picture, those are really long distance (>180 or even 270), and they'd make the plot rather messy.

![travel_time_body_waves1](https://cloud.githubusercontent.com/assets/302469/6263930/507be91a-b7ec-11e4-8b42-3f801494bc5f.png)

Note: You can see that `pPcP` is one of those phases that does not end at the right location.

QuLogic commented on misc/docs/source/tutorial/code_snippets/travel_time_body_waves.py:
```diff
@@ -0,0 +1,81 @@
+              horizontalalignment='center', verticalalignment='center',
+              bbox=dict(facecolor='white', edgecolor='none', alpha=0.7))
+ocr = (model.model.radiusOfEarth -
+       (model.model.sMod.vMod.iocbDepth + model.model.sMod.vMod.cmbDepth) / 2)
```
So this is kind of ugly. Maybe we want to add some shortcuts.

QuLogic commented on obspy/taup/tau.py:
```diff
@@ -11,14 +18,70 @@
+    "#cab2d6"]
+
+
+class SmartPolarText(matplotlib.text.Text):
```
Might want to make this private or put somewhere else.

QuLogic commented on obspy/taup/tau.py:
```diff
@@ -31,6 +94,177 @@ def __str__(self):
+        :type legend: bool or str
+        :param label_arrivals: Label the arrivals with their respective phase
+            names. This setting is only useful if you are plotting a single
+            phase as otherwise the names be large and possibly overlap or clip.
```
I missed a verb here; please add in 'may' or 'could' next time you update.

QuLogic commented on obspy/taup/tau.py:
```diff
@@ -31,6 +94,177 @@ def __str__(self):
+                    marker="*", color="#FEF215", markersize=20, zorder=10,
+                    markeredgewidth=1.5, markeredgecolor="0.3", clip_on=False)
+
+            # Pretty station marker.
```
FYI, I only change to annotations on the spherical side. The Cartesian wasn't going to run into the same problem.

ThomasLecocq commented:
just a note, the direct upward arrival "p" doesn't exist, right ? It's capital "P", then it can be named "pP" !?

claudiodsf commented:
@ThomasLecocq No, the direct upward P arrival is "p". "pP" is the arrival which first goes upwards and finally arrives at teleseismic distance as a P-wave. You can see a "pP" arrival at ~315° on @QuLogic plot.

ThomasLecocq commented:
humm... http://pasadena.wr.usgs.gov/office/hough/Storchak.pdf not sure about that...

claudiodsf commented:
At teleseismic distance, "P" is a mantle phase. I don't think that, at global scale, you can call a crustal P-wave just "P".  You can use though, like in the paper you're referring to, more specific notation like "Pg", "Pb" or "Pn".

krischer commented:
To clear up the confusion: The TauP phase name convention is a bit different from the IASPEI one. Its documented in the TauP manual and we should probably also document it in `obspy.taup`. 

http://www.seis.sc.edu/downloads/TauP/taup.pdf

ThomasLecocq commented:
> we should probably also document it in obspy.taup. 

yeah :-)

claudiodsf commented:
@QuLogic Why is that ray path green? What is your coloring criterion?

krischer commented:
@claudiodsf Its just green because there are two SKKS arrivals at the same distance.

claudiodsf commented:
@krischer Ah ok. I still think, though, that we should differentiate ray paths by colors, like in your first plots.

krischer commented:
Probably. Its still the case with the automatic plots.

The one from @qulogic is just to show off the taup interface in the tutorial and gallery :-) But we can still use different colors in there. I would also really like to distinguish P and S legs like in the famous Stein picture somewhere further up this thread. But that, as said, requires some changes further down the stack which will happen I guess.

QuLogic commented:
Yes, this example is just for the tutorial/gallery. The plotting code can't really handle multiple distances, so it is essentially many single plots over top of each other, and they end up the same colour. The green is the secondary arrival; there are secondary arrivals for `PPP` and `SSS`, but they bounce very near to other phase arrivals, so I found them confusing and left them out.

QuLogic commented on obspy/taup/tau.py:
```diff
@@ -11,14 +18,70 @@
+    :param distance: The requested distance.
     """
-    __slots__ = ["arrivals"]
+    __slots__ = ["model", "distance"]
```
So is it necessary to hold on to the distance here? Could you use `arrival.distance` vs. `arrival.purist_distance` instead? That way, the multiple arrival scenario in the tutorial could just be `+=` and a single `plot`.

mbyt commented:
Really, really nice plots! Now I finally understand what's taup good for :o).

krischer commented:
Travis and appveyor failures are unrelated. Can we merge this?

QuLogic commented:
The body wave example seems to have broken now.

QuLogic commented:
The only issue I have here is that the imports means you effectively need `matplotlib` installed correctly during installation. IIRC, depending on how you install, runtime dependencies might be installed after the package is built.
